Proper handling of both zero and null tx timeout

thelonelyvulpes opened the pull request:
Change tx timeout to allow users to pass 
- `null` (default server timeout).
- `TimeSpan.Zero` (No timeout).

thelonelyvulpes committed:
change to null

thelonelyvulpes committed:
update doc string

thelonelyvulpes committed:
remove inconsistent whitespace

thelonelyvulpes committed:
Unit Tests

thelonelyvulpes committed:
timeout in ms instead of min

robsdedude commented on Neo4j.Driver/Neo4j.Driver.Tests.TestBackend/Protocol/Session/SessionRun.cs:
```diff
@@ -25,32 +24,29 @@ public class SessionRunType
-            if (data.timeout != -1)
+            if (data.timeout.HasValue)
             {
-                var time = TimeSpan.FromMilliseconds(data.timeout);
-                configBuilder.WithTimeout(time);
+                configBuilder.WithTimeout(TimeSpan.FromMilliseconds(data.timeout.Value));
             }
```
I don't think this is doing what I envisioned it to do. There should be three cases covered here:
 * Set the timeout to `null` explicitly
 * Set the timeout to a given time span
 * Don't set the timeout at all

This code never runs into the first case.

thelonelyvulpes committed:
fix testcase knowing if timeout is implicit or explicit.

thelonelyvulpes committed:
5.0 change log for tracking

thelonelyvulpes committed:
fix not calling the converter

robsdedude committed:
Add tx timout 0 and null treatment to all relevant places

 * Session.BeginTransaction
 * Session.ReadTransaction
 * Session.WriteTransaction

Added re-writing of `ArgumentOutOfRangeException` raised on negative timeouts
to TestKit `DriverError`.

thelonelyvulpes committed:
add exception filter

thelonelyvulpes committed:
add param name to exception filter

thelonelyvulpes committed:
add filters

thelonelyvulpes committed:
remove tabs

thelonelyvulpes committed:
fix unit tests for zero and null cases

thelonelyvulpes committed:
refactoring SessionXType code to remove duplication

thelonelyvulpes committed:
Merge branch '5.0' into feature/allow-zero-and-null-tx-timeout
